nixos/modules/services/web-servers/caddy/default.nix

```diff
@@ -70,6 +70,28 @@ let
   configPath = "/etc/${etcConfigFile}";
 
   mkCertOwnershipAssertion = import ../../../security/acme/mk-cert-ownership-assertion.nix lib;
+
+  validateCaddyConfig =
+    adapter: configFile:
+    builtins.fromJSON (
+      builtins.readFile (
+        pkgs.runCommand "caddy-validate"
+          {
+            nativeBuildInputs = [ cfg.package ];
+          }
+          ''
+            if caddy validate --adapter ${toString adapter} --config ${configFile}; then
+              echo "true" > $out
+            else
+              echo "false" > $out
+            fi
+          ''
+      )
+    );
+
+  configValidationResult = validateCaddyConfig (
+    if cfg.adapter != null then cfg.adapter else "caddyfile"
+  ) cfg.configFile;
```

Comment:
Import From Derivation (IFD) is actively discouraged. I'd encourage you to implement this similar to `services.nginx` (through the use of `pkgs.writes.writeNginxConfig`) which writes the configuration file as part of the derivation, but fails to build if it is incorrent.